Add standalone visual object detection notebook. (#3586)
aws/amazon-sagemaker-examples: introduction_to_applying_machine_learning/visual_object_detection/README.md

This solution detects product defects with an end-to-end Deep Learning workflow for quality control in manufacturing process. The solution takes input of product images and identifies defect regions with bounding boxes. In particular, this solution uses a pre-trained Sagemaker object detection model and fine-tune on the target dataset.

This solution will demonstrate the immense advantage of fine-tuning a high-quality pre-trained model on the target dataset, both visually and numerically.

### Contents
1. [Overview](#overview)
   1. [How Does the Input Data Look Like?](#input)
   2. [How to Prepare Your Data to Feed into the Model?](#preparedata)
   3. [What are the Outputs?](#output)
   4. [What is the Estimated Cost?](#cost)
   5. [What Algorithms & Models are Used?](#algorithms)
   6. [How Does the Data Flow Look Like?](#dataflow)
2. [Solution Details](#solution)
   1. [Background](#background)
   2. [What is Visual Inspection?](#inspection)
   3. [What are the Problems?](#problems)
   4. [What Does this Solution Offer?](#offer)
3. [Architecture Overview](#architecture)
4. [Cleaning up](#cleaning-up)
5. [Customization](#customization)

## 1. Overview <a name="overview"></a>

> ### 1.1. How Does the Input Data Look Like? <a name="input"></a>

Author: fixed

> Input is an image of a defective / non-defective product. The training data should have relatively balanced classes, with annotations for ground truth defects (locations and defect types) per image. Here is an example used in the demo

Author: modified

> !["sample2"](https://sagemaker-solutions-prod-us-east-2.s3.us-east-2.amazonaws.com/sagemaker-defect-detection/docs/sample2.png)

Author: added some description

The NEU surface defect database (see [references](#references)) is a *balanced* dataset which contains

> Six kinds of typical surface defects of the hot-rolled steel strip are collected, i.e., rolled-in scale (RS), patches (Pa), crazing (Cr), pitted surface (PS), inclusion (In) and scratches (Sc). The database includes 1,800 grayscale images: 300 samples each of six different kinds of typical surface defects

> Here is a sample images of the six classes

Author: fixed

!["data sample"](https://sagemaker-solutions-prod-us-east-2.s3.us-east-2.amazonaws.com/sagemaker-defect-detection/docs/data.png)

### 1.2. How to Prepare Your Data to Feed into the Model? <a name="preparedata"></a>

There are data preparation and preprocessing steps and should be followed in the notebooks. It's critical to prepare your image annotations beforehand.

> For finetuning pretrained Sagemaker models, you need to prepare either a single `annotation.json` for all data, or a `RecordIO` file for both all images and all annotations. Check the notebook for details.

Author: There is only one notebook here.

### 1.3. What are the Outputs? <a name="output"></a>

> * For each image, the trained model will produce bounding boxes of detected visual defects (if any), the predicted defect type and prediction confidence score (0~1).

Author: fixed

* If you have a labeled test dataset, you could obtain the mean Average Precision (mAP) score for each model and compare among all the models.
   * For example, the mAP scores on a test set of the NEU dataset

     |     |  Type1 | Type1+HPO | Type2 | Type2+HPO|
     | --- |  --- | --- | --- | ---|
     | mAP | 0.067 | 0.226 | 0.371 | 0.375|

### 1.4. What is the Estimated Cost? <a name="cost"></a>

> * Running the notebook costs around $130~140 USD. This notebook provides advanced materials, including finetuning two types of pretrained Sagemaker models **till convergence**, with and without hyperparameter optimization (HPO), and result in four models for inference. You could choose to train either one model, or all four models according to your budget and requirements. The cost and runtime for training each model (assuming using p3.2xlarge EC2 instance) are:

Reviewer: We typically don't try to quote costs in docs because they're region-specific and change w/o notice. It's very nice to provide estimates though. So make sure you say, for example, or if the cost is x in y region, then here's how you'd do the math, rather than say this "z", is the cost... because you just don't know.

Author: Very nice suggestion. I added the "if the cost is x in y region"-like description. Thanks

   | Model | Cost (USD) | Runtime (Hours) | Billable time (Hours)|
   |:----------:|:---------------:|:----:|:-----:|
   |Type 1|    1.5     |       0.5       | 0.5|
   |Type 1 with HPO (20 jobs)|    30.6    |       1*        | 10|
   |Type 2|    4.6     |       1.5       | 1.5|
   |Type 2 with HPO (20 jobs)|     92     |       3*        | 30|
  (*) HPO tasks in this solution consider 20 jobs in total and 10 jobs in parallel. So 1 actual runtime hour amounts to 10 billable cost hours.

> * Please make sure you have read the cleaning up part [here](#cleaning-up) after training to avoid incurred cost from deployed models.

Reviewer: Please use a more descriptive link (not "here").

Author: fixed, replaced with "read the cleaning up part in [Section 4](#cleaning-up) after ..."

### 1.5. What Algorithms & Models are Used? <a name="algorithms"></a>

* The pretrained Sagemaker models include SSD models and FasterRCNN model, using either VGG, ResNet, or MobileNet as backbone, pretrained on either ImageNet, COCO, VOC, or FPN dataset.

> ### 1.6. How Does the Data Flow Look Like? <a name="dataflow"></a>

Reviewer: ```suggestion
### 1.6. What Does the Data Flow Look Like? <a name="dataflow"></a>
```

Author: fixed

![Data flow](https://sagemaker-solutions-prod-us-east-2.s3.us-east-2.amazonaws.com/sagemaker-defect-detection/docs/data_flow.png)

## 2. Solution Details <a name="solution"></a>

### 2.1. Background <a name="background"></a>

> According to the [Gartner](https://www.gartner.com/smarterwithgartner/gartner-top-10-strategic-technology-trends-for-2020/),  hyper-automation is the number one trend in 2020 and will continue advancing in future. When it comes to manufacturing, one of the main barriers to hyper-automation is in areas where Human involvements is still struggling to be reduced and intelligent systems have hard times to become on-par with Human visual recognition abilities and become mainstream, despite great advancement of Deep Learning in Computer Vision. This is mainly due to lack of enough annotated data (or when data is sparse) in areas such as _Quality Control_ sections where trained Human eyes still dominates.

Reviewer: ```suggestion
According to the [Gartner study on the top 10 strategic tech trends for 2020](https://www.gartner.com/smarterwithgartner/gartner-top-10-strategic-technology-trends-for-2020/),  hyper-automation is the number one trend in 2020 and will continue advancing in future. When it comes to manufacturing, one of the main barriers to hyper-automation is in areas where Human involvements is still struggling to be reduced and intelligent systems have hard times to become on-par with Human visual recognition abilities and become mainstream, despite great advancement of Deep Learning in Computer Vision. This is mainly due to lack of enough annotated data (or when data is sparse) in areas such as _Quality Control_ sections where trained Human eyes still dominates.
```

Author: fixed, thanks

### 2.2. What is Visual Inspection? <a name="inspection"></a>

> The **analysis of products on the production line for the purpose of Quality Control**. Visual inspection can also be used for internal and external assessment of the various equipment in a production facility such as storage tanks, pressure vessels, piping, and other equipment ([source](https://nanonets.com/blog/ai-visual-inspection/)) which expands to many industries from Electronics, Medical, Food and Raw Materials.

Reviewer: ```suggestion
The **analysis of products on the production line for the purpose of Quality Control**. Visual inspection can also be used for ([internal and external assessment of the various equipment in a production facility such as storage tanks, pressure vessels, piping, and other equipment](https://nanonets.com/blog/ai-visual-inspection/)) which expands to many industries from Electronics, Medical, Food and Raw Materials.
```

Author: I changed to the following:

The **analysis of products on the production line for the purpose of Quality Control**. According to [Everything you need to know about Visual Inspection with AI
](https://nanonets.com/blog/ai-visual-inspection/), visual inspection can also be used for internal and external assessment of the various equipment in a production facility such as storage tanks, pressure vessels, piping, and other equipment which expands to many industries from Electronics, Medical, Food and Raw Materials.

### 2.3. What are the Problems? <a name="problems"></a>

> * *Human visual inspection error* is a major factor in this area. According to this [report](https://www.osti.gov/servlets/purl/1476816)

Reviewer: More descriptive link would be good here.

Author: changed to * *Human visual inspection error* is a major factor in this area. According to the report [The Role of Visual Inspection in the 21st Century](https://www.osti.gov/servlets/purl/1476816)

   > Most inspection tasks are much more complex and typically exhibit error rates of 20% to 30% (Drury & Fox, 1975)

which directly translates to *cost*.

> * Cost: according to some [estimate](https://www.glassdoor.co.in/Salaries/us-quality-control-inspector-salary-SRCH_IL.0,2_IN1_KO3,28.htm), a trained quality inspector salary varies between 26K (US) - 60K per year.

Reviewer: More descriptive link would be good here.

Author: changed to "according to [glassdoor estimate](...)"

### 2.4. What Does this Solution Offer?  <a name="offer"></a>

This solution offers a complete solution using high-quality pretrained Sagemaker models to finetune on the target dataset with and without hyperparameter optimization (HPO).

The **most important** information this solution delivers, is that training a deep learning model from scratch on a small dataset can be both time-consuming and less effective, whereas finetuning a high-quality pretrained model, which was trained on large-scale dataset, could be both cost- and runtime-efficient and highly performant. Here are the sample detection results

<img src="patches_116.jpeg" alt="drawing" width="1200"/>

## 3. Architecture Overview <a name="architecture"></a>

> Here is architecture for the end-to-end training and deployment process

Reviewer: ```suggestion
The following illustration is the architecture for the end-to-end training and deployment process:
```

Author: fixed
